### PR TITLE
Security block. First part. Add User entity

### PR DESCRIPTION
**Pull request by yurlis**
Add the required DTOs and mapper. Add AuthenticationController. Do data validation in the new DTO classes. Add FieldMatch annotation for match password and repeatPassword fields. Consider to handle the RegistrationException using the GlobalExceptionHandler

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -144,4 +144,4 @@
 		</plugins>
 	</build>
 
-</project>
+</project>
```
Comment (MAMentorFX) [inline pom.xml]:
```suggestion
</project>

```

File: src/main/java/com/bookstoreapp/validator/isbn/IsbnCustomValidator.java [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 
 import jakarta.validation.ConstraintValidator;
 import jakarta.validation.ConstraintValidatorContext;
 
 import java.util.regex.Pattern;
 
-public class IsbnValidator implements ConstraintValidator<Isbn, String> {
+public class IsbnCustomValidator implements ConstraintValidator<IsbnConstraint, String> {
     private static final String PATTERN_OF_ISBN = "^(?=(?:\\D*\\d){10}(?:(?:\\D*\\d){3})?$)[\\d-]+$";
     /*
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/validator/isbn/IsbnCustomValidator.java]: Remove comments please

File: src/main/java/com/bookstoreapp/service/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
+        User savedUser = userRepository.save(user);
+        return userMapper.toUserResponse(savedUser);
+    }
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/UserServiceImpl.java]: multiple calls to DB. Make it trascational

File: src/main/java/com/bookstoreapp/dto/user/UserRegistrationRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+@FieldMatch.List({
+        @FieldMatch(first = "password", second = "repeatPassword",
+                message = "The password fields must match")
+})
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Email
+    private String email;
+    @NotBlank
+//    @Length(min = 8, max = 35, message = "Password must be between 5 and 35 characters")
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/dto/user/UserRegistrationRequestDto.java]: remove comments

File: src/main/java/com/bookstoreapp/exception/RegistrationException.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.bookstoreapp.exception;
+
+public class RegistrationException extends Exception {
+    public RegistrationException(String message) {
+        super(message);
+    }
+}
+
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/exception/RegistrationException.java]:
```suggestion
```

File: src/main/java/com/bookstoreapp/mapper/UserMapper.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.bookstoreapp.config.MapperConfig;
+import com.bookstoreapp.dto.user.UserRegistrationResponseDto;
+import com.bookstoreapp.controller.model.User;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+    UserRegistrationResponseDto toUserResponse(User user);
+}
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/mapper/UserMapper.java]:
```suggestion
}

```

File: src/main/java/com/bookstoreapp/repository/user/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.bookstoreapp.controller.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    Optional<User> findByEmail(String email);
+}
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/repository/user/UserRepository.java]:
```suggestion
}

```

File: src/main/java/com/bookstoreapp/repository/user/UserRepository.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.bookstoreapp.repository.user;
+
+import com.bookstoreapp.controller.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/repository/user/UserRepository.java]:
```suggestion
```

File: src/main/java/com/bookstoreapp/service/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can't register user");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/UserServiceImpl.java]: Use @MapStruct

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookstoreapp/validator/password/FieldMatchValidator.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        Object firstObj = new BeanWrapperImpl(value).getPropertyValue(firstFieldName);
+        Object secondObj = new BeanWrapperImpl(value).getPropertyValue(secondFieldName);
+        if (firstObj != null) {
+            return firstObj.equals(secondObj);
+        } else {
+            return secondObj == null;
+        }
```
Comment (Rommelua) [inline src/main/java/com/bookstoreapp/validator/password/FieldMatchValidator.java]:
```suggestion
            return Objects.equals(firstObj, secondObj);
```

File: src/main/java/com/bookstoreapp/validator/password/PasswordCustomValidator.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+    @Override
+    public void initialize(PasswordConstraint password) {
+    }
+
+    @Override
+    public boolean isValid(String password,
+                           ConstraintValidatorContext cxt) {
+        return password != null && password.matches(PASSWORD_REGEXP)
+                && (password.length() >= 8) && (password.length() < 128);
```
Comment (Rommelua) [inline src/main/java/com/bookstoreapp/validator/password/PasswordCustomValidator.java]: magic numbers

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookstoreapp/mapper/UserMapper.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.bookstoreapp.config.MapperConfig;
+import com.bookstoreapp.dto.user.UserRegistrationRequestDto;
+import com.bookstoreapp.dto.user.UserRegistrationResponseDto;
+import com.bookstoreapp.model.User;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+    UserRegistrationResponseDto toUserResponse(User user);
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/mapper/UserMapper.java]: In the future you may have more Dtos, not only Registration Response, so such naming is not clear, which Dto it returns

File: src/main/java/com/bookstoreapp/service/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Override
+    @Transactional
+    public UserRegistrationResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can't register user");
+        }
+        return userMapper.toUserResponse(userRepository.save(userMapper.toUserModel(requestDto)));
+    }
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/service/UserServiceImpl.java]: You dont need to use toUserModel here. You have to use return value of findByEmail, so change isPresent to orElseThrow.

File: src/main/java/com/bookstoreapp/validator/password/FieldMatchValidator.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        firstFieldName = constraintAnnotation.first();
+        secondFieldName = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        Object firstObj = new BeanWrapperImpl(value).getPropertyValue(firstFieldName);
+        Object secondObj = new BeanWrapperImpl(value).getPropertyValue(secondFieldName);
```
Comment (MAMentorFX) [inline src/main/java/com/bookstoreapp/validator/password/FieldMatchValidator.java]: Does it actually require creation of two different BeanWrapperImpl objects?

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**